controlplane/kubeadm/internal/workload_cluster_rbac.go

```diff
@@ -0,0 +1,150 @@
+	if err := w.Client.Create(ctx, obj); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "unable to create object on workload cluster")
+		}
+
+		if err := w.Client.Update(ctx, obj); err != nil {
+			return errors.Wrap(err, "unable to update object on workload cluster")
+		}
+	}
+	return nil
+}
+
+// AllowBootstrapTokensToGetNodes creates RBAC rules to allow Node Bootstrap Tokens to list nodes
+func (w *Workload) AllowBootstrapTokensToGetNodes(ctx context.Context) error {
+	if err := w.CreateOrUpdateObject(ctx, &rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetNodesClusterRoleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Rules: []rbac.PolicyRule{
+			{
+				Verbs:     []string{"get"},
+				APIGroups: []string{""},
+				Resources: []string{"nodes"},
+			},
+		},
+	}); err != nil {
+		return err
+	}
+
+	return w.CreateOrUpdateObject(ctx, &rbac.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetNodesClusterRoleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "ClusterRole",
+			Name:     GetNodesClusterRoleName,
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind: rbac.GroupKind,
+				Name: NodeBootstrapTokenAuthGroup,
+			},
+		},
+	})
+}
+
+func generateKubeletConfigMapName(version semver.Version) string {
+	return fmt.Sprintf(KubeletConfigMapRoleName, version.Major, version.Minor)
+}
+
+// ReconcileKubeletRBACBinding will create a RoleBinding for the new kubelet version during upgrades.
+// If the role binding already exists this function is a no-op.
+func (w *Workload) ReconcileKubeletRBACBinding(ctx context.Context, version semver.Version) error {
+	roleName := generateKubeletConfigMapName(version)
+
+	return w.CreateOrUpdateObject(ctx, &rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				APIGroup: rbac.GroupName,
+				Kind:     rbac.GroupKind,
+				Name:     NodesGroup,
+			},
+			{
+				APIGroup: rbac.GroupName,
+				Kind:     rbac.GroupKind,
+				Name:     NodeBootstrapTokenAuthGroup,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "Role",
+			Name:     roleName,
+		},
+	})
+
+}
+
+// ReconcileKubeletRBACRole will create a Role for the new kubelet version during upgrades.
+// If the role already exists this function is a no-op.
+func (w *Workload) ReconcileKubeletRBACRole(ctx context.Context, version semver.Version) error {
+	roleName := generateKubeletConfigMapName(version)
+	return w.CreateOrUpdateObject(ctx, &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:         []string{"get"},
+				APIGroups:     []string{""},
+				Resources:     []string{"configmaps"},
+				ResourceNames: []string{roleName},
```

Comment:
Can't use roleName here (kubeadm:kubelet-config-x.y). Have to use the resource name without the kubeadm: prefix.

Reply:
The kubeadm code has `configMapRBACName()` for the role name and `kubeadmconstants.GetKubeletConfigMapName()` for the resource name. They look very similar but they're different.

Reply:
After much head scratching I found out the same about exactly the same time as you commented :(